File: CountingBits.java
//https://leetcode.com/problems/counting-bits/

class Solution {
    public int[] countBits(int n) {
        int ans[] = new int[n+1];
        ans[0] = 0;
        int oldIndex = 1;
        int newCountPoint = 1;
        for(int i = 1; i < n+1; i++){     
            if(i==newCountPoint){
                ans[i] = 1;
                oldIndex = 1;
                newCountPoint*=2;
            }else{
                ans[i] = 1+ans[oldIndex];
                oldIndex++;
            }

        }
        return ans;
    }
}

//0       0
//1       1

//10      2
//11      3

//100     4
//101     5
//110     6
//111     7

//1000    8 
//1001    9
//1010    10
//1011    11
//1100    12
//1101    13
//1110    14
//1111    15

//10000   16
//10001   17
//10010   18
//10011   19
//10100   20
//10101   21
//10110   22
//10111   23
//11000   24
//11001   25
//11010   26
//11011   27
//11100   28
//11101   29
//11110   30
//11111   31

File: Easy/CountingBits/CountingBits.py
# https://leetcode.com/problems/counting-bits/description/

class Solution:
    def countBits(self, n: int) -> List[int]:
        result = []
        rangeSize = 2
        for i in range(0,n+1):
            if i < 2:
                result.append(i)
            else:
                if i==rangeSize*2:
                    rangeSize*=2
                result.append(result[i-rangeSize]+1)

        return result

        
        # 0                    0
        # 1                    1

        # 10                   1
        # 11                   2

        # 100                  1
        # 101                  2
        # 110                  2
        # 111                  3

        # 1000                 1
        # 1001                 2
        # 1010                 2
        # 1011                 3
        # 1100                 2
        # 1101                 3
        # 1110                 3
        # 1111                 4

        # 10000                1
        # 10001                2
        # 10010                2
        # 10011                3

        # 10100                2
        # 10101                3
        # 10110                3
        # 10111                4